crates/shared/src/bad_token/cache.rs

```diff
@@ -53,38 +59,43 @@ impl CachingDetector {
         }
     }
 
-    fn insert_into_cache(&self, token: H160, quality: TokenQuality) {
+    async fn insert_into_cache(&self, token: H160, quality: TokenQuality) {
         self.cache
-            .lock()
-            .unwrap()
+            .write()
+            .await
             .insert(token, (Instant::now(), quality));
     }
 
-    fn insert_many_into_cache(&self, tokens: impl Iterator<Item = (H160, TokenQuality)>) {
-        let mut cache = self.cache.lock().unwrap();
+    async fn insert_many_into_cache(&self, tokens: impl Iterator<Item = (H160, TokenQuality)>) {
+        let mut cache = self.cache.write().await;
         let now = Instant::now();
-        for (token, quality) in tokens {
+        tokens.into_iter().for_each(|(token, quality)| {
             cache.insert(token, (now, quality));
-        }
+        });
     }
 
     fn spawn_maintenance_task(self: Arc<Self>) {
-        let cache_expiry = self.cache_expiry;
-        let maintenance_timeout = cache_expiry.div(10).max(Duration::from_secs(60));
+        // We need to prefetch the token quality the `prefetch_time` before the cache
+        // expires
+        let prefetch_time_to_expire = self.cache_expiry - self.prefetch_time;
+        // The maintenance period has to be at least double of the prefetch time in
+        // order to guarantee that the prefetch time is executed before the token
+        // quality expires. This is because of the Nyquist–Shannon sampling theorem.
```

Comment:
> Nyquist–Shannon sampling theorem

You like to throw around these fancy words, don't you? 🤓

Reply:
I like to show I am an educated person :nerd_face: :rofl: